### PR TITLE
Create Abstract classes and Tetrahedron

### PR DESCRIPTION
**Pull request by novalain**
Restructured classes and Scene. Only implemented Tetrahedron so far, creating Sphere as separate issue

### CODE REVIEW

File: src/main.cc [hunk 1]
```diff
@@ -5,6 +5,8 @@
 #include "triangle.h"
 #include "ray.h"
 #include "scene.h"
+#include "triangle.h"
```
Comment (Carlbaum) [inline src/main.cc]: @ row #8: 2nd include of triangle.h

**Review by Carlbaum — CHANGES_REQUESTED** [review]
Remove row #8 from main.cc: 
triangle.h included twice

File: include/tetrahedron.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#ifndef TETRAHEDRON_H
+#define TETRAHEDRON_H
+
+#include "triangle_based_object.h"
+#include "triangle.h"
+
+class Tetrahedron : TriangleBasedObject {
+  public:
```
Comment (Carlbaum) [inline include/tetrahedron.h]: No indents on public, private nor protected in .h files

Comment (Carlbaum) [reply include/tetrahedron.h]: @ row #8

File: include/triangle_based_object.h [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#ifndef TRIANGLE_BASED_OBJECT_H
+#define TRIANGLE_BASED_OBJECT_H
+
+#include "scene_object.h"
+#include "triangle.h"
+#include <vector>
+
+class TriangleBasedObject : SceneObject {
+  protected:
```
Comment (Carlbaum) [inline include/triangle_based_object.h]: @ row #9 no indent

Comment (novalain) [reply include/triangle_based_object.h]: Already fixed in later commit 😌  View all changes in all commits in this PR by clicking any commit message and "View all"
skarmavbild 2016-09-17 kl 02 05 47

File: include/tetrahedron.h (hunk 2 above)
Comment (novalain) [reply include/tetrahedron.h]: also fixed already

File: include/triangle.h [hunk 4]
```diff
@@ -21,7 +21,7 @@ class Triangle
   glm::vec3 get_normal() { return normal_; }
   glm::vec3 get_color() { return color_; }
 
-  bool RayIntersection(Ray* ray);
+  bool RayIntersection(Ray& ray);
```
Comment (Carlbaum) [inline include/triangle.h:24]: Wasn't this changed to return a float.. not a bool?

**Review by Carlbaum — APPROVED** [review]

File: include/triangle.h (hunk 4 above)
Comment (novalain) [reply include/triangle.h:24]: Yes but in the other PR. will rebase and resolve conflicts :sleepy:

**Pull request closed without merging**